### ISSUE TITLE
split the update method

### ISSUE THREAD

**Issue by eyelash** [google/xi-editor#95]
As not all fields of the update method are always set (if you send an undo, you'll get an update without scrollto for example) I propose to split the update method into 4 methods for height, lines, scrollto and cursors.

For example like this:

```
update_height 12
```

```
update_lines {
    "first_line": 0,
    "lines":[["hello",["sel",4,5]]]
}
```

```
update_scrollto [0,4]
```

```
update_cursors [[0,4]]
```

**Comment by eyelash** [comment]
The current design is also not optimal for multiple cursors (the front-end would need to store the cursors per line to know which ones are added and which are removed) so what about just splitting out the cursors from the update?

**Comment by raphlinus** [comment]
I think what I'd like to do for that is send the line's cursors for each line which has a change of cursors. So it'd look something like `"cursors": [{"line": 3, "cursors": []}, {"line": 4, "cursors": [1]}]` (that's something you'd see when moving the cursor from line 3 to 4). Each of the elements of height, lines, cursors, scrollto would be optional. Would that meet your needs?

Splitting off cursors into a separate method still has the problem I mentioned above; the front-end would draw with stale cursor info (the cursor might not even address a valid location), then draw correctly.

**Comment by eyelash** [comment]
This would be a field in the update method, right? So why not just `"cursors": [[4,1]]` and only send the cursors that are inside the visible area? Or why separate them per line? The front-end needs to draw and therefore store them separately anyway (to support blinking and stuff).

**Comment by raphlinus** [comment]
Yes, a field in the update method.

The core doesn't know the exact visible area, it's just a guess. On Cocoa, the UI toolkit prerenders well outside the window, to make scrolling smoother.

I think you're right though that storing them per-line is over-complicating things. Probably best to just send all the cursors every time they change.

**Comment by eyelash** [comment]
I just realized my idea wouldn't work so well with the render_lines call so per lines cursors is probably still the best approach. Sending all the cursors means the front-end would need to decide which cursors are visible and I'm not sure how well this would scale with lots of cursors (for example searching for something common inside a large file and inserting cursors for each result).

**Comment by eyelash** [comment]
So then what about sending each line like this:
```json
{
	"index":3,
	"text":["hello",["sel",1,3],["fg",2,4,255]],
	"cursors":[]
}
```
Either text or cursors would be optional and the lines field of update would be an array of such lines. This way you could also send one update if for example only lines 3 and 5 changed.

**Comment by DivineDominion** [comment]
Short note: To rule out the possibility of dealing with duplicate data, I'd make the line number the key and the changes values in a dictionary:

```
{ # ...
  3: {
    "text":["hello",["sel",1,3],["fg",2,4,255]],
    "cursors":[]
  }
}
```

I have to familiarize myself with the RPC protocol a lot more. But I 1 aggregated update model that may or may not hold text and/or cursor changes truly sounds superior to atomic changes that may not even be in the correct order. Then the front-end can apply every change from the update as it deems fit and afterwards trigger a re-draw.

**Comment by eyelash** [comment]
I think this can be closed now that #156 has landed and the remaining work is captured in #231 and #317.
